docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,322 @@
+#### Prometheus-metrics-adapter precludes the use of sources of custom metrics
+With current K8s design, only one API extension can serve custom metrics. Per this GEP, the custom metrics extension
+point is occupied by prometheus-metrics-adapter.
+
+#### Heavyweight Prometheus infrastructure feeds signal back into cluster control loop
+The proposed signal flow brings two Prometheus instances (shoot prometheus and seed aggregate prometheus) into the shoot
+cluster control loop. This has major implications on reliability. The alternative is a small, custom-built component
+registered as seed custom metrics API extension, which directly calls the ShootKapi metrics endpoints and outputs
+request rate.
+
+The current approach was chosen solely because it leverages an existing open source component and avoids authoring a
+custom one. The historical record of Gardener Prometheus components reliability is deemed acceptable for the purpose.  
+
+The proposed complex signal path introduces additional latency to scaling response. At present, the resulting scaling
+latency is deemed satisfactory.
+
+While the scope of this GEP is limited to using the Prometheus signal path, the proposed design allows for an easy
+substitution with a signal path based on aforementioned simple custom component alternative. This supports future
+changes based on accumulated field experience, or evolving solution needs. 
+
+#### Observed fluctuations in actual compute efficiency affecting ShootKapi workloads
+There is a known source of horizontal scaling inefficiency, which presents an opportunity for substantial further
+improvement. It is not addressed by this proposal and instead deferred to a separate implementation stage, as it
+requires a strictly incremental effort on top of this GEP.
+
+An approximate 50% reduction (factor 0.5) of the effective compute power per unit of CPU was occasionally observed on
+some productive ShootKapi instances. The change occurs with a quick cutoff transition, and has been observed for periods
+or up to 2 hours. This variance currently forces substantial safety margin upon HPA's target metric value, effectively
+necessitating more replicas to be deployed than required during usual operation.
+
+The source is believed to be competition for compute resources with sibling workloads, either on the K8s, or
+infrastructure level, but the circumstances have not been conductive to investigation after the fact.
+A subsequent stage, outside the scope of this GEP is planned to focus on understanding the root cause, and further
+optimizing HPA's target metric value accordingly.
+
+Once this biggest source of inefficiency has been resolved, yet another round of scaling performance fine-tuning
+is planned, where instead of overall request rate, HPA will be driven by a weighted average, reflecting the different
+compute cost of different request categories (e.g. a cluster-scoped LIST is more cpu intensive than a resource-scoped PUT)
+
+## Alternatives
```

Comment:
This might sound naive, but why not scaling up horizontally if there are too many `429 too many requests` errors  at the ShootKapi instances of a shoot?

Reply:
This sounds like the correct thing to do: detect an overload situation as indicated by HTTP 429 responses and react to it by adding more instances. However, I think there are two drawbacks to this approach:
* If a single client (for whatever reason) hits only a single replica, causing the HTTP 429 responses, this approach will continue to scale up and the overload situation will still exist. I think if we're adding autoscaling in any way we have to ensure that our approach is guaranteed to improve the metric we're evaluating.
* There might be other indicators of overload as well, for example response times getting slower, HTTP 503 responses, or similar things. I think currently we are not sure that we understand the whole spectrum of how kube-apiserver indicates overload that we should react to by scaling up.

Reply:
At least to me, it does not sound naive, and I'd like to make use of the smarter algorithm built into API priority and fairness (and surfacing as `HTTP 429` errors).

To your question: In practice, because historically issues have been with resource exhaustion, and not request level throttling (or I'm oblivious of it). And in load tests we didn't see in-flight requests reaching anywhere near current limits. Also, since Gardener does not have API priority and fairness deployed yet, the exact behaviour of HTTP-429s is seen as something that's about to change.

In theory (and if we're looking at HTTP-429 as complete replacement of request count as driver), because HTTP-429s don't reflect memory starvation. Also, they provide a cue when to scale out, but not when to scale back in.

I think that once we have API priority and fairness, HTTP-429s could make a valuable supplement to the current metric, catching cases of CPU starvation which are missed by that metric, and potentially enabling us to use narrower safety margin.

Reply:
Marking this resolved to keep the PR manageable. Please don't hesitate to reopen, if you'd like to discuss further.